### PR TITLE
235 polish delegation contract

### PR DESCRIPTION
**Pull request by R-Santev**

### CODE REVIEW

File: contracts/HydraDelegation/DelegationPoolLib.sol [hunk 1]
```diff
@@ -159,19 +209,152 @@ library DelegationPoolLib {
...
+        uint256 paramsIndex
+    ) private view returns (uint256 balance, int256 correction) {
+        if (paramsIndex >= pool.delegatorsParamsHistory[delegator].length) {
+            revert DelegateRequirement({src: "vesting", msg: "INVALID_PARAMS_INDEX"});
+        }
+
+        DelegationPoolDelegatorParams memory params = pool.delegatorsParamsHistory[delegator][paramsIndex];
+        if (params.epochNum > epochNumber) {
+            revert DelegateRequirement({src: "vesting", msg: "LATE_BALANCE_CHANGE"});
+        } else if (params.epochNum == epochNumber) {
```
Comment (SamBorisov) [inline contracts/HydraDelegation/DelegationPoolLib.sol:295]: Can we optimize this to have less if statements? else if (params.epochNum != epochNumber && paramsIndex != pool.delegatorsParamsHistory[delegator].length - 1) { do the logic needed}

File: contracts/HydraDelegation/DelegationPoolLib.sol [hunk 2]
```diff
@@ -159,19 +209,152 @@ library DelegationPoolLib {
...
+        reward = claimableRewards(pool, account, epochNumber, balanceChangeIndex);
         pool.claimedRewards[account] += reward;
     }
+
+    /**
+     * @notice Deletes  the historical data of a delegator
+     * @param pool the DelegationPool to clean the historical data from
+     * @param account the address of the  delegator to clean the historical data from
+     */
+    function cleanDelegatorHistoricalData(DelegationPool storage pool, address account) internal {
```
Comment (SamBorisov) [inline contracts/HydraDelegation/DelegationPoolLib.sol:232]: Would it be good idea to clean the data when swapping, for the new validator

**Review by SamBorisov — CHANGES_REQUESTED** [review]

File: contracts/HydraDelegation/DelegationPoolLib.sol (hunk 1 above)
Comment (R-Santev) [reply contracts/HydraDelegation/DelegationPoolLib.sol:295]: Good catch

File: contracts/HydraDelegation/DelegationPoolLib.sol (hunk 2 above)
Comment (R-Santev) [reply contracts/HydraDelegation/DelegationPoolLib.sol:232]: GC!

**Comment by SamBorisov** [comment]
forge tests do not pass

**Review by SamBorisov — APPROVED** [review]

**Pull request merged**